File: python/password_breaker.py
import hasher

class password_breaker:
    def __init__(self, chars, digests, max_output_length = 50):
        self.chars = chars
        self.passwords_to_break = digests
        self.max_output_length = max_output_length
        self.broken_passwords = {}

    def start_breaking(self):
        self.produce_string()
        return self.broken_passwords

    def produce_string(self, word = ""):
        if(len(word) >= self.max_output_length) or len(self.passwords_to_break) == 0:
            return

        new_word = ""

        for i in range (0, len(self.chars)):
            new_word = word + self.chars[i]
            sha256_hash = hasher.make_hash_sha256(new_word)
            print(new_word + " : " + sha256_hash)
            self.compare_digests(sha256_hash,new_word)
            if len(self.passwords_to_break) == 0:
                return
            else:
                self.produce_string(new_word)

        return

    def compare_digests(self, digest, word):
        for password_digest in self.passwords_to_break:
            if password_digest == digest:
                print("HIT!")
                print("Digest: " + digest)
                print("Word  : " + word)
                self.broken_passwords[digest] = word
                self.passwords_to_break.remove(password_digest)

File: dotnet/PasswordBreaker.cs
using System.Collections.Generic;
using System.Linq;

namespace dotnet
{
    public class PasswordBreaker
    {
        private string chars;
        private List<string> passwordsToBreak;
        private int maxOutputLength;

        private Dictionary<string, string> brokenPasswords;
        private Hasher hasher;

        public PasswordBreaker(string charsToCombine, IEnumerable<string> digests, int maxLength = 50)
        {
            chars = charsToCombine;
            passwordsToBreak = digests.ToList();
            maxOutputLength = maxLength;

            brokenPasswords = new Dictionary<string, string>();
            hasher = new Hasher();
        }

        public IDictionary<string, string> StartBreaking()
        {
            ProduceString("");
            return brokenPasswords;
        }

        private void ProduceString(string word)
        {
            if(word.Length >= maxOutputLength || passwordsToBreak.Count == 0)
            {
                return;
            }

            string newWord = "";
            string digest = "";
            for (int i = 0; i < chars.Length; i++)
            {
                newWord = word + chars[i];
                digest = hasher.Make_Hash_SHA256(newWord);
                CompareDigests(digest, newWord);
                if(passwordsToBreak.Count == 0)
                {
                    return;
                }
                else
                {
                    ProduceString(newWord);
                }
            }
        }

        private void CompareDigests(string digest, string word)
        {
            foreach (var item in passwordsToBreak)
            {
                if(digest.Equals(item))
                {
                    System.Console.WriteLine("HIT!");
                    System.Console.WriteLine($"Digest: {digest}");
                    System.Console.WriteLine($"Word  : {word}");
                    brokenPasswords[digest] = word;
                    passwordsToBreak.Remove(digest);
                    CompareDigests(digest, word);
                    return;
                }
            }
        }
    }
}